CMakeLists.txt

```diff
@@ -0,0 +1,178 @@
+file(MAKE_DIRECTORY ${CMAKE_LIBRARY_OUTPUT_DIRECTORY})
+file(MAKE_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
+
+cuda_add_library(algorithm SHARED
+        query/algorithm.hpp
+        query/algorithm.cu
+        query/binder.hpp
+        query/dimension_transform.cu
+        query/filter.cu
+        query/functor.cu
+        query/geo_intersects.cu
+        query/hash_lookup.cu
+        query/hll.cu
+        query/iterator.hpp
+        query/measure_transform.cu
+        query/scratch_space_transform.cu
+        query/sort_reduce.cu
+        query/transform.cu
+        query/transform.hpp
+        query/utils.cu
+        query/utils.hpp)
+
+
+################################
+# GTest
+################################
+set(GTEST_ROOT $ENV{GTEST_ROOT})
+link_directories(${GTEST_ROOT})
+enable_testing()
+################################
+# Unit Tests
+################################
+file(GLOB QUERY_UNITTEST_FILES query/*_unittest.cu)
+cuda_add_executable(all_unittest ${QUERY_UNITTEST_FILES})
+target_include_directories(all_unittest PRIVATE ${GTEST_ROOT}/include ${gtest_SOURCE_DIR})
+#add_dependencies(all_unittest algorithm)
+# Link test executable against gtest & gtest_main
+target_link_libraries(all_unittest gtest gtest_main algorithm pthread)
+add_test(all_unittest ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/all_unittest)
+
+###############################
+# golang
+###############################
+
+execute_process(COMMAND bash -c "find . -name '*.go' | grep -v \
+  -e vendor -e go-build \
+  -e build \
+  -e '.*/\\..*' \
+  -e '.*/_.*' \
+  -e '.*/mocks.*'"
+        OUTPUT_VARIABLE ALL_GO_SRC)
+string(REPLACE "\n" " " ALL_GO_SRC ${ALL_GO_SRC})
+
+execute_process(
+        COMMAND bash -c "git diff origin/master HEAD --name-only --diff-filter=AMR | grep -e \".*\\.go\""
+        OUTPUT_VARIABLE CHANGED_GO_SRC
+        )
+if (CHANGED_GO_SRC)
+    string(REPLACE "\n" " " CHANGED_GO_SRC ${CHANGED_GO_SRC})
+endif()
+
+
+execute_process(COMMAND bash -c "find .  -type f \
+    \\( -iname \\*.cu -o -iname \\*.h -o -iname \\*.hpp -o -iname \\*.c \\) \
+     | grep -v \
+  -e vendor -e go-build \
+  -e build \
+  -e '.*/\\..*' \
+  -e '.*/_.*' \
+  -e '.*/mocks.*'"
+        OUTPUT_VARIABLE ALL_C_SRC)
+string(REPLACE "\n" " " ALL_C_SRC ${ALL_C_SRC})
+
+execute_process(
+        COMMAND bash -c "git diff HEAD origin/master --name-only | grep -e \".*\\.\\(cu\\|c\\|h\\|hpp\\)\""
+        OUTPUT_VARIABLE CHANGED_C_SRC
+        )
+if (CHANGED_C_SRC)
+    string(REPLACE "\n" " " CHANGED_C_SRC ${CHANGED_C_SRC})
+endif()
+
+add_custom_target(lint-all
+        COMMAND ./scripts/clang-lint.sh ${ALL_C_SRC}
+        COMMAND ./scripts/golang-lint.sh ${ALL_GO_SRC}
+        VERBATIM)
+
+add_custom_target(lint
+        COMMAND ./scripts/clang-lint.sh ${CHANGED_C_SRC}
+        COMMAND ./scripts/golang-lint.sh ${CHANGED_GO_SRC}
+        VERBATIM)
+
+add_custom_target(golang-dep DEPENDS vendor/glide.updated)
```

Comment:
is it possible merge this two lines with `add_custom_target` with its own commands?

Reply:
no, have to be one target and one commands and the command to generate the fileshttps://stackoverflow.com/questions/7862300/how-to-make-cmake-target-executed-whether-specified-file-was-changed

Reply:
I see, good to know